network/hybridNetwork.go

```diff
@@ -0,0 +1,233 @@
+	if err != nil {
+		return nil, err
+	}
+	wsnet, err := NewWebsocketNetwork(log, cfg, phonebookAddresses, genesisID, networkID, nodeInfo, p2pnet.PeerID(), p2pnet.PeerIDSigner())
+	if err != nil {
+		return nil, err
+	}
+	return &HybridP2PNetwork{
+		p2pNetwork: p2pnet,
+		wsNetwork:  wsnet,
+		genesisID:  genesisID,
+	}, nil
+}
+
+// Address implements GossipNode
+func (n *HybridP2PNetwork) Address() (string, bool) {
+	// TODO map from configuration? used for REST API, goal status, algod.net, etc
+	if n.useP2PAddress {
+		return n.p2pNetwork.Address()
+	}
+	return n.wsNetwork.Address()
+}
+
+type hybridNetworkError struct{ p2pErr, wsErr error }
+
+func (e *hybridNetworkError) Error() string {
+	return fmt.Sprintf("p2pErr: %s, wsErr: %s", e.p2pErr, e.wsErr)
+}
+func (e *hybridNetworkError) Unwrap() []error { return []error{e.p2pErr, e.wsErr} }
+
+func (n *HybridP2PNetwork) runParallel(fn func(net GossipNode) error) error {
+	var wg sync.WaitGroup
+	var p2pErr, wsErr error
+
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		p2pErr = fn(n.p2pNetwork)
+	}()
+	go func() {
+		defer wg.Done()
+		wsErr = fn(n.wsNetwork)
+	}()
+	wg.Wait()
+
+	if p2pErr != nil && wsErr != nil {
+		return &hybridNetworkError{p2pErr, wsErr}
+	}
+	if p2pErr != nil {
+		return p2pErr
+	}
+	if wsErr != nil {
+		return wsErr
+	}
+	return nil
+}
+
+// Broadcast implements GossipNode
+func (n *HybridP2PNetwork) Broadcast(ctx context.Context, tag protocol.Tag, data []byte, wait bool, except Peer) error {
+	return n.runParallel(func(net GossipNode) error {
+		return net.Broadcast(ctx, tag, data, wait, except)
+	})
+}
+
+// Relay implements GossipNode
+func (n *HybridP2PNetwork) Relay(ctx context.Context, tag protocol.Tag, data []byte, wait bool, except Peer) error {
+	return n.runParallel(func(net GossipNode) error {
+		return net.Relay(ctx, tag, data, wait, except)
+	})
+}
+
+// Disconnect implements GossipNode
+func (n *HybridP2PNetwork) Disconnect(badnode DisconnectablePeer) {
+	net := badnode.GetNetwork()
+	if net == n.p2pNetwork {
+		n.p2pNetwork.Disconnect(badnode)
+	} else if net == n.wsNetwork {
+		n.wsNetwork.Disconnect(badnode)
+	} else {
+		panic("badnode.GetNetwork() returned a network that is not part of this HybridP2PNetwork")
+	}
+}
+
+// DisconnectPeers implements GossipNode
+func (n *HybridP2PNetwork) DisconnectPeers() {
+	_ = n.runParallel(func(net GossipNode) error {
+		net.DisconnectPeers()
+		return nil
+	})
+}
+
+// RegisterHTTPHandler implements GossipNode
+func (n *HybridP2PNetwork) RegisterHTTPHandler(path string, handler http.Handler) {
+	n.p2pNetwork.RegisterHTTPHandler(path, handler)
+	n.wsNetwork.RegisterHTTPHandler(path, handler)
+}
+
+// RequestConnectOutgoing implements GossipNode
+func (n *HybridP2PNetwork) RequestConnectOutgoing(replace bool, quit <-chan struct{}) {}
+
+// GetPeers implements GossipNode
+func (n *HybridP2PNetwork) GetPeers(options ...PeerOption) []Peer {
```

Comment:
P2P peers and WS peers are different structs, so would not be treated as dups representing the same node.

Reply:
we have dedup logic in now in https://github.com/algorand/go-algorand/pull/6035 so should be fine